### ISSUE TITLE
Unexpected behavior of HttpClientHandler.ServerCertificateCustomValidationCallback

### ISSUE THREAD

**Issue by dhuyvett** [dotnet/corefx#33677]
I have a dotnet core 2.1 application that is doing an explicit DNS lookup before using HttpClient to call the URL with the IP address substituted for the hostname.

As expected, for https URLS this results in 

    System.Security.Authentication.AuthenticationException : The remote certificate is invalid according to the validation procedure.

I add the original hostname to the request in the Host header, and the exception persists.

I add a trivial ServerCertificateCustomValidationCallback . and the exception is no longer thrown.

    httpHandler.ServerCertificateCustomValidationCallback = (message, cert, chain, errors) => errors == SslPolicyErrors.None;

I remove the Host header from the request, or put bad data in that header, and AuthenticationException returns.

It appears that the very existence of ServerCertificateCustomValidationCallback changes the way that the HttpClientHandler validates the hostname/certificate match.

Is this dependable behavior?  Is there a better way in 2.1 to trigger host name validation to prefer the Host header? Is this documented anywhere?  (I haven't been able to find any discussion of the current behavior)

**Comment by davidsh** [comment]
@stephentoub @geoffkizer

**Comment by stephentoub** [comment]
@dhuyvett, what OS/version is this?

**Comment by stephentoub** [comment]
@davidsh, I expect this is more likely to be in SslStream than in SocketsHttpHandler.

**Comment by dhuyvett** [comment]
.NET Core SDK (reflecting any global.json):
     Version:   2.1.402
     Commit:    3599f217f4
    
    Runtime Environment:
     OS Name:     Mac OS X
     OS Version:  10.13
     OS Platform: Darwin
     RID:         osx.10.13-x64
     Base Path:   /usr/local/share/dotnet/sdk/2.1.402/
    
    Host (useful for support):
      Version: 2.1.4
      Commit:  85255dde3e

**Comment by stephentoub** [comment]
cc: @bartonjs

**Comment by geoffkizer** [comment]
> It appears that the very existence of ServerCertificateCustomValidationCallback changes the way that the HttpClientHandler validates the hostname/certificate match.

Yes, that's the point of the callback: you can substitute your own cert validation logic.

I'm not clear why you aren't just using the hostname in the URL.

**Comment by stephentoub** [comment]
I assume the concern in this case though is that the callback is called with errors as None?

What's the behavior on .NET Framework? What about .NET Core on Windows?

**Comment by davidsh** [comment]
>I'm not clear why you aren't just using the hostname in the URL.

It is not unreasonable to expect that the 'Host: ' request header is used for TLS/SSL certificate name match checking.  This has been a frequent customer request previously.  Many times, the URL is based on a load-balancer IP and the scenario requires the 'Host: ' name header to carry the name used for TLS/SSL certificate matching.

**Comment by dhuyvett** [comment]
If there is no user callback defined, corefx validates the host name in the cert is validated using the host from the request URL. (it ignores the Host header)

If a user callback is defined ... **even if all the user callback does is return whether errors were found by corefx** ... then corefx validates the cert using the host from the Host header.

To be clear: validation using the Host header is what I am after.  I had resigned myself to having to write my own ServerCertificateCustomValidationCallback to support this, But when I wrote the callback, I found it wasn't needed ... unless it wasn't there, in which case it was needed.

Either validation by URL host (boo!) or validation by Host header (yay!) are valid behavior.  But this behavior shouldn't change just because I set a callback to do additional processing.
